Add DatatypeDef

Aatch opened the pull request:
Add a DatatypeDef type that contains information for both structs and enums. Most places already treated structs like single-variant enums or vice-versa, so this actually simplifies a lot of code.

There is a small but significant performance increase in typechecking and translation (pre-optimisations) of about 5% in both phases. All the other phases seemed to have some performance improvements too, but the differences were too small to be meaningful.

This change is also a step towards overall simplification of the type system code. Given how many places are now treating enums and structs completely identically, a move towards a unified `ty_data` type is somewhat feasible.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTING.md](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

eddyb commented:
I would prefer a name like `DataType`, but that's... bikeshedding.
cc @nikomatsakis @nick29581

eddyb commented on src/librustc/lint/builtin.rs:
```diff
@@ -542,7 +542,11 @@ impl LintPass for BoxPointers {
+        let ty = match ty::expr_ty_opt(cx.tcx, e) {
+            Some(ty) => ty,
+            None => cx.tcx.sess.span_bug(e.span, "No type for expression"),
+        };
```
Doesn't/can't `ty::expr_ty` handle this?

Aatch replied:
Ah right, that reminds me, I was going to fold the `span_bug` stuff into `ty::expr_ty`. Makes debugging easier.

eddyb commented on src/librustc/middle/mem_categorization.rs:
```diff
@@ -1205,13 +1205,16 @@ impl<'t,'tcx,TYPER:Typer<'tcx>> MemCategorizationContext<'t,TYPER> {
+            let enum_def = ty::lookup_datatype_def(self.tcx(), enum_did);
+            // Univariant enums don't need a downcast
+            if !enum_def.is_univariant() {
+                self.cat_downcast(pat, cmt.clone(), cmt.ty, variant_did)
```
Down... cast? This is pre-existing, but what does it actually mean? Some sort of ancient refinement relic?

eddyb commented on src/librustc/middle/traits/coherence.rs:
```diff
@@ -178,9 +178,9 @@ fn ty_is_local_constructor<'tcx>(tcx: &ty::ctxt<'tcx>, ty: Ty<'tcx>) -> bool {
-            def_id.krate == ast::LOCAL_CRATE
+        ty::ty_enum(def, _) |
+        ty::ty_struct(def, _) => {
+            def.def_id.krate == ast::LOCAL_CRATE
```
Makes me wonder if we should have an `is_local()` method on `DefId` so we can do `def.def_id.is_local()`.

eddyb commented on src/librustc/middle/ty.rs:
```diff
@@ -629,6 +616,7 @@ pub struct CtxtArenas<'tcx> {
     substs: TypedArena<Substs<'tcx>>,
     bare_fn: TypedArena<BareFnTy<'tcx>>,
     region: TypedArena<Region>,
+    def: TypedArena<DatatypeDef<'tcx>>,
```
Could use a more descriptive name (I would confuse it with `middle::def::Def`).

eddyb commented on src/librustc/middle/ty.rs:
```diff
@@ -2301,6 +2289,128 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
+    pub name: Name,
+    pub disr_val: Disr,
+    pub fields: Vec<FieldTy<'tcx>>,
+    pub vis: Visibility,
```
I thought `enum` variants are never private.

eddyb commented on src/librustc/middle/ty.rs:
```diff
@@ -2301,6 +2289,128 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
+    pub name: Name,
+    pub vis: ast::Visibility,
+    pub origin: DefId,
+    pub ty: Cell<Ty<'tcx>>
```
This needs a comment explaining why `Cell` is used here.
Similarly, the distinction between `id` and `origin` is not obvious, could also use a comment.

eddyb commented on src/librustc/middle/ty.rs:
```diff
@@ -2301,6 +2289,128 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
+    }
+}
+
+impl<'tcx> fmt::Debug for DatatypeDef<'tcx> {
```
Shouldn't this be a `Repr` implementation?

Aatch replied:
I added the `Debug` impl because `sty` derives `Debug` meaning that `DatatypeDef` need to impl `Debug` too.

eddyb commented on src/librustc/middle/ty.rs:
```diff
@@ -3181,21 +3295,20 @@ pub fn sequence_element_type<'tcx>(cx: &ctxt<'tcx>, ty: Ty<'tcx>) -> Ty<'tcx> {
 }
 
-pub fn simd_type<'tcx>(cx: &ctxt<'tcx>, ty: Ty<'tcx>) -> Ty<'tcx> {
+pub fn simd_type<'tcx>(_cx: &ctxt<'tcx>, ty: Ty<'tcx>) -> Ty<'tcx> {
```
This and a few other functions might benefit from not taking the unused `tcx` argument at all.

eddyb commented on src/llvm:
```diff
@@ -1 +1 @@
-Subproject commit 4891e6382e3e8aa89d530aa18427836428c47157
```
Rebase failure (run `make` and let it update submodules and/or avoid committing everything).

eddyb commented on src/librustc/middle/ty.rs:
```diff
@@ -2301,6 +2289,128 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
+
+    pub fn is_tuple_variant(&self) -> bool {
+        !self.fields.is_empty() &&
+            self.fields.iter().all(|f| f.name == token::special_names::unnamed_field)
```
I would avoid comparing to  `unnamed_field` outside of an `is_unnamed` method on `FieldTy`.

eddyb commented on src/librustc_trans/trans/adt.rs:
```diff
@@ -373,10 +376,10 @@ fn find_discr_field_candidate<'tcx>(tcx: &ty::ctxt<'tcx>,
+            let fields = &def.variants[0].fields[];
             for (j, field) in fields.iter().enumerate() {
-                let field_ty = ty::lookup_field_type(tcx, def_id, field.id, substs);
+                let field_ty = field.ty.get().subst(tcx, substs);
```
Maybe `.ty.get()` would be better off as `.ty()` - can the `ty` field be fully private?
Would be nice to know one cannot simply `field.ty.set(ty)`.

Aatch replied:
Good point, you'd still need a `set_ty` method or similar, but I could always do a check to make sure that the type already set on the field is `ty_err` (meaning that you can only set it from `ty_err` to something else).

eddyb commented on src/librustc_trans/trans/intrinsic.rs:
```diff
@@ -364,8 +364,8 @@ pub fn trans_intrinsic_call<'a, 'blk, 'tcx>(mut bcx: Block<'blk, 'tcx>,
-            if !return_type_is_void(ccx, tp_ty) {
-                // Just zero out the stack slot. (See comment on base::memzero for explanation)
+            if !return_type_is_void(tp_ty) {
+                // Just zero out the stack slot. (See comment on base::memzero for explaination)
```
Typo? Why was this comment changed in the first place?

Aatch replied:
Ah, there was a conflict and I missed that difference.

eddyb commented:
LGTM modulo some nits, but I'd be happy with a second pair of eyes.

alexcrichton commented:
r? @nikomatsakis 

(I'm probably not the right reviewer for this)

nikomatsakis commented:
Sorry for the delay. I somehow missed this PR.

nikomatsakis commented on src/librustc/middle/ty.rs:
```diff
@@ -2301,6 +2289,177 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
+
+    pub fn ty(&self) -> Ty<'tcx> {
+        self.ty.get()
+    }
```
The current setup, iiuc, initially stores a ty_err and then updates it to the actual type. This seems like it will hide bugs where we read the type of a field before it is fully initialized. I would prefer to use `Cell<Option<Ty<'tcx>>>` and initially set it to `None`. Then `ty` can call `self.ty.get().unwrap()`, which will give the same (stronger) guarantees we get today. Similarly `set_ty` can check that the type is set exactly once. (There is no space inefficiency because `Ty<'tcx>` is a non-nullable-pointer.)

nikomatsakis commented:
ok, I have to run for the moment. I'm a big fan of this general approach. I've been wanting to see the struct/enum code made nicer (and unified!) for a long time, and I think moving away from storing all the information in distinct side-tables is generally a good idea. I've only had time though to read over the high-level details of what is here, so I'll want some more time to soak in finer points.

nikomatsakis commented on src/librustc_privacy/lib.rs:
```diff
@@ -897,21 +898,17 @@ impl<'a, 'tcx, 'v> Visitor<'v> for PrivacyVisitor<'a, 'tcx> {
-                    ty::ty_struct(ctor_id, _) => {
-                        // RFC 736: ensure all unmentioned fields are visible.
-                        // Rather than computing the set of unmentioned fields
-                        // (i.e. `all_fields - fields`), just check them all.
```
Nit: Please don't drop this comment.

nikomatsakis commented on src/rust-installer:
```diff
@@ -1 +1 @@
-Subproject commit 60fd8abfcae50629a3fc664bd809238fed039617
+Subproject commit e577c97b494be2815b215e3042207d6d4b7c5516
```
this subproject commit is presumably unintentional?

nikomatsakis commented:
I think I'd like to see the `Cell<Option<Ty<'tcx>>>` replaced with an `Ivar` type in `librustc/middle/util`:

```rust
/// A "write once" cell.
pub struct Ivar<T:Copy> {
    data: Cell<Option<T>>
}

impl<T:Copy> Ivar<T> {
    fn get(&self) -> Option<T> {
        self.data.get()
    }

    fn fulfill(&self, value: T) {
        assert!(self.data.get().is_none());
        self.data.set(value);
    }

    fn unwrap(&self) -> T {
        self.get().unwrap()
    }
}
```

This has the advantage that (a) privacy ensures that we will only write to the cell oncce and (b) it can, in the future, be made `Sync` (we'd have to use atomic writes).

Aatch commented:
I've squashed all the previous commits since rebasing was getting difficult due to functions being removed/replaced and me having to manually re-remove them post-rebase.

I have also added the `Ivar` type and used it for `DatatypeDef`

nikomatsakis commented on src/librustc/metadata/encoder.rs:
```diff
@@ -1142,56 +1133,57 @@ fn encode_info_for_item(ecx: &EncodeContext,
-        encode_struct_fields(rbml_w, &fields[..], def_id);
+          /* Encode def_ids for each field and method
+          for methods, write all the stuff get_trait_method
+          needs to know*/
```
Pre-existing,but this comment formatting is terrible.

nikomatsakis commented:
@aatch I commented on a few points that were made before but not corrected in your latest rebase. Overall I'm pretty happy with this.

One thing I was wondering -- perhaps it would make sense to merge `ty_struct` and `ty_enum` into `ty_data` or something like that. Just something to think about, for now.

nikomatsakis commented:
Modulo those nits though I'm ready to r+.

pnkfelix commented on src/librustc/util/ivar.rs:
```diff
@@ -0,0 +1,41 @@
+    data: Cell<Option<T>>
+}
+
+impl<T:Copy> Ivar<T> {
```
I personally know what an "I-var" is, but it might be nice to either have some comments here just saying what the big picture is about them, or a pointer to http://en.wikipedia.org/wiki/Futures_and_promises or something.

Aatch commented:
@nikomatsakis yeah, I looked at merging them into a `ty_data`, but it's a non-trivial change on top of this already fairly extensive change. Definitely something to do in the near future though.

nikomatsakis commented:
r+ from me, but of course there tidy errors it looks like

bors commented:
:umbrella: The latest upstream changes (presumably #22801) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
There's a lot of `&x[]` slicing syntax, that I believe to be deprecated.

nikomatsakis commented:
I was thinking about `Ivar` and I remembered that as I originally envisioned it, `Ivar` wouldn't require `T:Copy` but, rather, would take advantage of the fact that because it is only set once, it can return an `&T` to the contents even without using a `RefCell` (this would require unsafe code, of course). This doesn't have to go in the PR, I'm just noting it for future reference. ;)

eddyb commented:
@nikomatsakis sounds like an "OnceCell".

pnkfelix commented:
@eddyb I am not 100% sure I would apply the `Cell` terminology here, since part of the point is that these things do not start with an initial value, while all of the current kinds of cells do start with an initial value, right?

eddyb commented:
@pnkfelix fair enough, I was just trying to force it into `std::cell`, but it probably fits better somewhere in `std::sync`.

nikomatsakis commented:
@Aatch do you plan to rebase this?

nikomatsakis commented:
@Aatch would you object if *I* rebased this?

Aatch commented:
@nikomatsakis go for it. Sorry, I've been really busy with work + life recently, not much time for compiler-hacking.

nikomatsakis commented:
@Aatch I started it. Lots of conflicts (of course). I might just try to cherry-pick and re-implement instead...gotta finish a few other things first.

alexcrichton commented:
Closing in favor of @nikomatsakis's upcoming rebase
